src/linear_algebra/matrix.lean

```diff
@@ -1026,3 +1028,121 @@ lemma matrix.dot_product_eq_zero_iff {v : n → R} : (∀ w, matrix.dot_product
 ⟨λ h, matrix.dot_product_eq_zero v h, λ h w, h.symm ▸ zero_dot_product w⟩
 
 end
+
+namespace matrix
+
+variables {m n : Type u} [decidable_eq n] [fintype n] [decidable_eq m] [fintype m] {R : Type v}
+  [comm_ring R]
+
+lemma to_block_matrix_det (M : matrix m m R) (p : m → Prop) [decidable_pred p] :
+  M.det = (matrix.from_blocks (to_block M p p) (to_block M p (λ j, ¬p j))
+    (to_block M (λ j, ¬p j) p) (to_block M (λ j, ¬p j) (λ j, ¬p j))).det :=
+begin
+  rw ← matrix.det_reindex_self (equiv.sum_compl p).symm M,
+  unfold det,
+  congr, ext σ, congr, ext,
+  generalize hy : σ x = y,
+  cases x; cases y;
+  simp only [matrix.reindex_apply, to_block_apply, equiv.symm_symm,
+    equiv.sum_compl_apply_inr, equiv.sum_compl_apply_inl,
+    from_blocks_apply₁₁, from_blocks_apply₁₂, from_blocks_apply₂₁, from_blocks_apply₂₂],
+end
+
+lemma to_square_block_det (M : matrix m m R) (b : m → ℕ) (k : ℕ) :
+  (to_square_block M b k).det = (to_square_block' M (λ i, b i = k)).det := by simp
+
+lemma upper_two_block_triangular_det' (M : matrix m m R) (p : m → Prop) [decidable_pred p]
+  (h : ∀ i (h1 : ¬p i) j (h2 : p j), M i j = 0) :
+  M.det = (to_square_block' M p).det * (to_square_block' M (λ i, ¬p i)).det :=
+begin
+  rw to_block_matrix_det M p,
+  convert upper_two_block_triangular_det (to_block M p p) (to_block M p (λ j, ¬p j))
+    (to_block M (λ j, ¬p j) (λ j, ¬p j)),
+  ext,
+  exact h ↑i i.2 ↑j j.2
+end
+
+lemma equiv_block_det (M : matrix m m R) {p q : m → Prop} [decidable_pred p] [decidable_pred q]
+  (e : ∀x, q x ↔ p x) : (to_square_block' M p).det = (to_square_block' M q).det :=
+by convert matrix.det_reindex_self (equiv.subtype_equiv_right e) (to_square_block' M q)
+
+/-- Let `b` map rows and columns of a square matrix `M` to `n` blocks. Then
+  `upper_block_triangular_matrix M n b` says the matrix is upper block triangular. -/
+def upper_block_triangular_matrix {o : Type*} [fintype o] (M : matrix o o R) (n : ℕ)
+  (b : o → ℕ) := (∀ i, b i < n) ∧ (∀ i j, b j < b i → M i j = 0)
```

Comment:
Is this easier or harder to work with as follows?
```suggestion
  (b : o → fin n) := ∀ i j, b j < b i → M i j = 0
```

Reply:
I originally tried to work with `fin n`, but it turns out to be much simpler to have `b` as I have it now. Also, I realized that the current `upper_block_triangular_matrix` isn't really a definition of "upper". Simply changing `b` to `(λ i, n - 1 - b i)` gives you "lower" already. I'll change things around to not use "upper" to make this clearer. I really think it is easier to use also in the form it is stated now.

Reply:
I've added lemmas using the `b : o → fin n` definition (but proved them using the `b : o → nat`) definition.